manual/CHAPTER_TextRtlil.tex

```diff
@@ -0,0 +1,252 @@
+
+The grammar is also not meant to represent the exact grammar used in the RTLIL frontend, since that grammar is specific to processing by lex and yacc, and is somewhat less understandable than simple EBNF notation.
+
+\section{Lexical elements}
+
+\subsection{Identifiers}
+
+There are three types of identifiers in RTLIL:
+
+\begin{itemize}
+    \item Publically visible identifiers
+    \item Auto-generated identifiers
+    \item Dotted numeric identifiers
+\end{itemize}
+
+\begin{indentgrammar}{<autogen-id>}
+<id> ::= <public-id> | <autogen-id> | <dotted-id>
+
+<public-id> ::= "\textbackslash" <nonws>$+$
+
+<autogen-id> ::= "\textdollar" <nonws>$+$
+
+<dotted-id> ::= "." <decimal-digit>$+$
+\end{indentgrammar}
+
+\subsection{Values}
+
+A \textit{value} consists of a width in bits and a bit representation, most significant bit first. Bits may be any of:
+\begin{itemize}
+    \item \texttt{0}: A logic zero value
+    \item \texttt{1}: A logic one value
+    \item \texttt{x}: An unknown logic value (or don't care in case patterns)
+    \item \texttt{z}: A high-impedance value (or don't care in case patterns)
+    \item \texttt{m}: A marked bit (internal use only)
+    \item \texttt{-}: A don't care value
+\end{itemize}
+
+An \textit{integer} is simply a signed integer value in decimal format.
+
+\begin{indentgrammar}{<binary-digit>}
+<value> ::= <decimal-digit>$+$ \texttt{\textbf{'}} <binary-digit>$*$
+
+<binary-digit> ::= "0" | "1" | "x" | "z" | "m" | "-"
+
+<integer> ::= "-"$?$ <decimal-digit>$+$
+\end{indentgrammar}
+
+\subsection{Strings}
+
+A string is a series of characters delimited by double-quote characters. Within a string, certain escapes can be used:
+
+\begin{itemize}
+    \item \texttt{\textbackslash n}: A newline
+    \item \texttt{\textbackslash t}: A tab
+    \item \texttt{\textbackslash \textit{ooo}}: A character specified as a one, two, or three digit octal value
+\end{itemize}
+
+All other characters may be escaped by a backslash, and become the following character. Thus:
+
+\begin{itemize}
+    \item \texttt{\textbackslash \textbackslash}: A backslash
+    \item \texttt{\textbackslash \"}: A double-quote
+    \item \texttt{\textbackslash r}: An 'r' character
+\end{itemize}
+
+\subsection{Comments}
+
+A comment starts with a \texttt{\textbf{\#}} character and proceeds to the end of the line. All comments are ignored.
+
+\section{File}
+
+A file consists of zero or more designs. A design may be a module, an attribute statement, or an autoindex statement.
+
+Note that in general, statements are terminated by an end of line.
+
+\begin{indentgrammar}{<design>}
+<file> ::= <design>$*$
+
+<design> ::= <module> | <attr-stmt> | <autoidx-stmt>
+\end{indentgrammar}
+
+\subsection{Modules}
+
+A module consists of zero or more module statements.
+
+\begin{indentgrammar}{<module-stmt>}
+<module> ::= "module" <id> <eol> <module-stmt>$*$ "end" <eol>
```

Comment:
I said it before, but attributes really ought to be attached to the module here.

Reply:
I can make that more explicit using the grammar, although those auto-index statements, whose function I don't know, appear to be able to be interspersed with module attributes and modules. In any case, I've also elaborated a bit on what attributes apply to in the attributes section.

Reply:
For example, a possible refactor that might make this clearer is:

`<file> ::= ((<autoidx-stmt> | <attr-stmt>)* <module>)*`

Reply:
> those auto-index statements, whose function I don't know

Practically, an `autoidx` statement should always appear only at the very start of an RTLIL file (if at all), and only once. This isn't enforced but perhaps should be.

The function of them is simple. Sometimes Yosys needs to generate an unique name. It picks e.g. `$flatten$N`. The `N` part is the value of the `autoidx` global, which is subsequently incremented. This global has to be dumped into RTLIL, otherwise e.g. dumping and running a pass would have different properties than just running a pass on a warm design.

Reply:
Ah, ok, I can document that. I could also again refactor the grammar so that autoidx only appears at the beginning of the file?

```
<file> ::= <autoidx-stmt>? <module>*
<module> ::= <module-attr-stmt>* <module-stmt> <module-body> <module-end-stmt>
<module-attr-stmt> := <attr-stmt>
```

Reply:
Now you got it! Although I'd probably use something like...
```
<file> ::= <autoidx-stmt>? <module>*
<module> ::= <attr-stmt>* <module-stmt> <module-body> <module-end-stmt>
```
since attributes are all the same in RTLIL no matter what they are attached to.

Reply:
Done!